homeassistant/components/dsmr/sensor.py

```diff
@@ -942,54 +972,85 @@ def __init__(
+            telegram, self._mbus_id, self.entity_description.obis_reference
         )
-        if dsmr_object is None:
-            return None
 
-        # Get the attribute value if the object has it
-        attr: str | None = getattr(dsmr_object, attribute)
-        return attr
+        if dsmr_object and dsmr_object.value is not None:
+            if self._is_averaged_sensor:
+                try:
+                    self._raw_values.append(float(dsmr_object.value))
+                except (ValueError, TypeError):
+                    LOGGER.debug(
+                        "Could not convert %s value %s to float for averaging",
+                        self.entity_description.key,
+                        dsmr_object.value,
+                    )
+            # For non-power sensors, we don't need to accumulate,
+            # calculate_value will use the latest _telegram.
 
-    @property
-    def available(self) -> bool:
-        """Entity is only available if there is a telegram."""
-        return self.telegram is not None
+    @callback
+    def calculate_value(self) -> None:
+        """Calculate the value to report based on accumulated data."""
+        if self._telegram is None:
+            self._value = None
+            self._raw_values = []  # Ensure clear on disconnect
+            return
 
-    @property
-    def native_value(self) -> StateType:
-        """Return the state of sensor, if available, translate if needed."""
-        value: StateType
-        if (value := self.get_dsmr_object_attr("value")) is None:
-            return None
+        dsmr_object = get_dsmr_object(
+            self._telegram, self._mbus_id, self.entity_description.obis_reference
+        )
+
+        if self._is_averaged_sensor:
+            if self._raw_values:
+                avg_value = statistics.mean(self._raw_values)
+                self._value = round(avg_value, DEFAULT_PRECISION)
+            # If no raw values, keep the previous value unless disconnected (handled above)
+            self._raw_values = []  # Always clear after processing
+            return
+
+        # Logic for non-power sensors (use latest value)
+        if dsmr_object is None or dsmr_object.value is None:
+            self._value = None
+            return
+
+        value = dsmr_object.value
 
         if self.entity_description.obis_reference == "ELECTRICITY_ACTIVE_TARIFF":
-            return self.translate_tariff(value, self._entry.data[CONF_DSMR_VERSION])
+            self._value = self.translate_tariff(value, self._dsmr_version)
+            return
 
-        with suppress(TypeError):
+        with suppress(TypeError, ValueError):
             value = round(float(value), DEFAULT_PRECISION)
 
         # Make sure we do not return a zero value for an energy sensor
         if not value and self.state_class == SensorStateClass.TOTAL_INCREASING:
-            return None
+            self._value = None
+        else:
+            self._value = value
```

Comment:
It feels like we should have a function that "consumes" accumulated values. So it still could be a return value here.